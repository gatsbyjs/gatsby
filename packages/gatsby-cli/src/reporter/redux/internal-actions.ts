import { uuid } from "gatsby-core-utils"
import signalExit from "signal-exit"
import { Dispatch } from "redux"

import { getStore } from "./"
import {
  Actions,
  ActivityLogLevels,
  ActivityStatuses,
  ActivityTypes,
} from "../constants"
import {
  IPendingActivity,
  ICreateLog,
  ISetStatus,
  IStartActivity,
  ICancelActivity,
  IEndActivity,
  IUpdateActivity,
  IActivityErrored,
  IGatsbyCLIState,
  ISetLogs,
  IRenderPageTree,
} from "./types"
import {
  delayedCall,
  getActivity,
  getElapsedTimeMS,
  getGlobalStatus,
} from "./utils"
import { IStructuredError, ErrorCategory } from "../../structured-errors/types"
import { IRenderPageArgs } from "../types"

const ActivityStatusToLogLevel = {
  [ActivityStatuses.Interrupted]: ActivityLogLevels.Interrupted,
  [ActivityStatuses.Failed]: ActivityLogLevels.Failed,
  [ActivityStatuses.Success]: ActivityLogLevels.Success,
}

let weShouldExit = false
signalExit(() => {
  weShouldExit = true
})

let cancelDelayedSetStatus: (() => void) | null

let pendingStatus: ActivityStatuses | "" = ``

// We debounce "done" statuses because activities don't always overlap
// and there is timing window after one activity ends and before next one starts
// where technically we are "done" (all activities are done).
// We don't want to emit multiple SET_STATUS events that would toggle between
// IN_PROGRESS and SUCCESS/FAILED in short succession in those cases.
export const setStatus =
  (status: ActivityStatuses | "", force: boolean = false) =>
  (dispatch: Dispatch<ISetStatus>): void => {
    const currentStatus = getStore().getState().logs.status

    if (cancelDelayedSetStatus) {
      cancelDelayedSetStatus()
      cancelDelayedSetStatus = null
    }

    if (
      status !== currentStatus &&
      (status === ActivityStatuses.InProgress || force || weShouldExit)
    ) {
      dispatch({
        type: Actions.SetStatus,
        payload: status,
      })
      pendingStatus = ``
    } else {
      // use pending status if truthy, fallback to current status if we don't have pending status
      const pendingOrCurrentStatus = pendingStatus || currentStatus

      if (status !== pendingOrCurrentStatus) {
        pendingStatus = status
        cancelDelayedSetStatus = delayedCall(() => {
          setStatus(status, true)(dispatch)
        }, 1000)
      }
    }
  }

export const createLog = ({
  level,
  text,
  statusText,
  duration,
  group,
  code,
  type,
  category,
  filePath,
  location,
  docsUrl,
  context,
  activity_current,
  activity_total,
  activity_type,
  activity_uuid,
  stack,
  pluginName,
}: {
  level: string
  text?: string
  statusText?: string
  duration?: number
  group?: string
  code?: string
  type?: string
  category?: keyof typeof ErrorCategory
  filePath?: string
  location?: IStructuredError["location"]
  docsUrl?: string
  context?: string
  activity_current?: number
  activity_total?: number
  activity_type?: string
  activity_uuid?: string
  stack?: IStructuredError["stack"]
  pluginName?: string
}): ICreateLog => {
  return {
    type: Actions.Log,
    payload: {
      level,
      text: !text ? `\u2800` : text,
      statusText,
      duration,
      group,
      code,
      type,
      category,
      filePath,
      location,
      docsUrl,
      context,
      activity_current,
      activity_total,
      activity_type,
      activity_uuid,
      timestamp: new Date().toJSON(),
      stack,
      pluginName,
    },
  }
}

type ActionsToEmit = Array<IPendingActivity | ReturnType<typeof setStatus>>
export const createPendingActivity = ({
  id,
  status = ActivityStatuses.NotStarted,
}: {
  id: string
  status?: ActivityStatuses
}): ActionsToEmit => {
  const globalStatus = getGlobalStatus(id, status)
  return [
    setStatus(globalStatus),
    {
      type: Actions.PendingActivity,
      payload: {
        id,
        type: ActivityTypes.Pending,
        startTime: process.hrtime(),
        status,
      },
    },
  ]
}

type QueuedStartActivityActions = Array<
  IStartActivity | ReturnType<typeof setStatus>
>
export const startActivity = ({
  id,
  text,
  type,
  status = ActivityStatuses.InProgress,
  current,
  total,
}: {
  id: string
  text: string
  type: ActivityTypes
  status?: ActivityStatuses
  current?: number
  total?: number
}): QueuedStartActivityActions => {
  const globalStatus = getGlobalStatus(id, status)

  return [
    setStatus(globalStatus),
    {
      type: Actions.StartActivity,
      payload: {
        id,
        uuid: uuid.v4(),
        text,
        type,
        status,
        startTime: process.hrtime(),
        statusText: ``,
        current,
        total,
      },
    },
  ]
}

type QueuedEndActivity = Array<
  ICancelActivity | IEndActivity | ICreateLog | ReturnType<typeof setStatus>
>

export const endActivity = ({
  id,
  status,
}: {
  id: string
  status: ActivityStatuses
}): QueuedEndActivity | null => {
  const activity = getActivity(id)
  if (!activity) {
    return null
  }

  const actionsToEmit: QueuedEndActivity = []
  const durationMS = getElapsedTimeMS(activity)
  const durationS = durationMS / 1000

  if (activity.type === ActivityTypes.Pending) {
    actionsToEmit.push({
      type: Actions.CancelActivity,
      payload: {
        id,
        status: ActivityStatuses.Cancelled,
        type: activity.type,
        duration: durationS,
      },
    })
  } else if (activity.status === ActivityStatuses.InProgress) {
    if (activity.errored) {
      status = ActivityStatuses.Failed
    }
    actionsToEmit.push({
      type: Actions.EndActivity,
      payload: {
        uuid: activity.uuid,
        id,
        status,
        duration: durationS,
        type: activity.type,
      },
    })

    if (activity.type !== ActivityTypes.Hidden) {
      actionsToEmit.push(
        createLog({
          text: activity.text,
          level: ActivityStatusToLogLevel[status],
          duration: durationS,
          statusText:
            activity.statusText ||
            (status === ActivityStatuses.Success &&
            activity.type === ActivityTypes.Progress
              ? `${activity.current}/${activity.total} ${(
                  (activity.total || 0) / durationS
                ).toFixed(2)}/s`
              : undefined),
          activity_uuid: activity.uuid,
          activity_current: activity.current,
          activity_total: activity.total,
          activity_type: activity.type,
        })
      )
    }
  }

  const globalStatus = getGlobalStatus(id, status)
  actionsToEmit.push(setStatus(globalStatus))

  return actionsToEmit
}

export const updateActivity = ({
  id = ``,
  ...rest
}: {
  id: string
  statusText?: string
  total?: number
  current?: number
}): IUpdateActivity | null => {
  const activity = getActivity(id)
  if (!activity) {
    return null
  }

  return {
    type: Actions.UpdateActivity,
    payload: {
      uuid: activity.uuid,
      id,
      ...rest,
    },
  }
}

export const setActivityErrored = ({
  id,
}: {
  id: string
}): IActivityErrored | null => {
  const activity = getActivity(id)
  if (!activity) {
    return null
  }

  return {
    type: Actions.ActivityErrored,
    payload: {
      id,
    },
  }
}

export const setActivityStatusText = ({
  id,
  statusText,
}: {
  id: string
  statusText: string
}): IUpdateActivity | null =>
  updateActivity({
    id,
    statusText,
  })

export const setActivityTotal = ({
  id,
  total,
}: {
  id: string
  total: number
}): IUpdateActivity | null =>
  updateActivity({
    id,
    total,
  })

export const activityTick = ({
  id,
  increment = 1,
}: {
  id: string
  increment: number
}): IUpdateActivity | null => {
  const activity = getActivity(id)
  if (!activity) {
    return null
  }

  return updateActivity({
    id,
    current: (activity.current || 0) + increment,
  })
}

export const setLogs = (logs: IGatsbyCLIState): ISetLogs => {
  return {
    type: Actions.SetLogs,
    payload: logs,
  }
}

export const renderPageTree = (payload: IRenderPageArgs): IRenderPageTree => {
  return {
    type: Actions.RenderPageTree,
    payload,
  }
}
