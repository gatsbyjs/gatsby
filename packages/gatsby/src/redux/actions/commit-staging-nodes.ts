import { ActionsUnion } from "../types"
import { internalCreateNodeWithoutValidation } from "./internal"
import { actions as publicActions } from "./public"
import { getNode } from "../../datastore"

import { store } from "../index"

export const commitStagingNodes = (
  transactionId: string
): Array<ActionsUnion> => {
  const transaction = store
    .getState()
    .nodesStaging.transactions.get(transactionId)
  if (!transaction) {
    return []
  }

  const actions: Array<ActionsUnion> = [
    {
      type: `COMMIT_STAGING_NODES`,
      payload: {
        transactionId,
      },
    },
  ]

  const nodesState = new Map()
  for (const action of transaction) {
    if (action.type === `CREATE_NODE_STAGING`) {
      nodesState.set(action.payload.id, action)
    } else if (action.type === `DELETE_NODE_STAGING` && action.payload?.id) {
      nodesState.set(action.payload.id, undefined)
    }
  }
  for (const [id, actionOrDelete] of nodesState.entries()) {
    if (actionOrDelete) {
      actions.push(
        ...internalCreateNodeWithoutValidation(
          actionOrDelete.payload,
          actionOrDelete.plugin,
          actionOrDelete
        )
      )
    } else {
      // delete case
      actions.push(publicActions.deleteNode(getNode(id)))
    }
  }

  return actions
}
