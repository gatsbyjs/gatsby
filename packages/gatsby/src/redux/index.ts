import {
  applyMiddleware,
  combineReducers,
  createStore,
  DeepPartial,
  Middleware,
  ReducersMapObject,
  Store,
} from "redux"
import _ from "lodash"

import { mett } from "../utils/mett"
import thunk, { ThunkMiddleware, ThunkAction, ThunkDispatch } from "redux-thunk"
import * as rawReducers from "./reducers"
import { writeToCache, readFromCache } from "./persist"
import { IGatsbyState, ActionsUnion, GatsbyStateKeys } from "./types"

const persistedReduxKeys = [
  `nodes`,
  `typeOwners`,
  `statefulSourcePlugins`,
  `status`,
  `components`,
  `jobsV2`,
  `staticQueryComponents`,
  `webpackCompilationHash`,
  `pageDataStats`,
  `pages`,
  `staticQueriesByTemplate`,
  `pendingPageDataWrites`,
  `queries`,
  `html`,
  `slices`,
  `slicesByTemplate`,
]

const reducers = persistedReduxKeys.reduce((acc, key) => {
  const rawReducer = rawReducers[key]
  acc[key] = function (state, action): any {
    if (action.type === `RESTORE_CACHE` && action.payload[key]) {
      return action.payload[key]
    } else {
      return rawReducer(state, action)
    }
  }

  return acc
}, rawReducers)

// Create event emitter for actions
export const emitter = mett()

// Read old node data from cache.
export const readState = (): IGatsbyState => {
  try {
    const state = readFromCache() as IGatsbyState
    if (state.nodes) {
      // re-create nodesByType
      state.nodesByType = new Map()
      state.nodes.forEach(node => {
        const { type } = node.internal
        if (!state.nodesByType.has(type)) {
          state.nodesByType.set(type, new Map())
        }
        // The `.has` and `.set` calls above make this safe
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        state.nodesByType.get(type)!.set(node.id, node)
      })
    }

    // jsonDataPaths was removed in the per-page-manifest
    // changes. Explicitly delete it here to cover case where user
    // runs gatsby the first time after upgrading.
    delete state[`jsonDataPaths`]

    return state
  } catch (e) {
    return {} as IGatsbyState
  }
}

export interface IMultiDispatch {
  <T extends ActionsUnion | ThunkAction<any, IGatsbyState, any, ActionsUnion>>(
    action: Array<T>
  ): Array<T>
}

/**
 * Redux middleware handling array of actions
 */
const multi: Middleware<IMultiDispatch> =
  ({ dispatch }) =>
  next =>
  (action: ActionsUnion): ActionsUnion | Array<ActionsUnion> =>
    Array.isArray(action) ? action.filter(Boolean).map(dispatch) : next(action)

export type GatsbyReduxStore = Store<IGatsbyState> & {
  dispatch: ThunkDispatch<IGatsbyState, any, ActionsUnion> & IMultiDispatch
}

export const configureStore = (
  initialState: IGatsbyState
): GatsbyReduxStore => {
  const store = createStore(
    combineReducers<IGatsbyState>({ ...reducers }),
    initialState,
    applyMiddleware(thunk as ThunkMiddleware<IGatsbyState, ActionsUnion>, multi)
  )

  store.dispatch({
    type: `INIT`,
  })

  return store
}

export const store: GatsbyReduxStore = configureStore(
  process.env.GATSBY_WORKER_POOL_WORKER ? ({} as IGatsbyState) : readState()
)

/**
 * Allows overloading some reducers (e.g. when setting a custom datastore)
 */
export function replaceReducer(
  customReducers: Partial<ReducersMapObject<IGatsbyState>>
): void {
  store.replaceReducer(
    combineReducers<IGatsbyState>({ ...reducers, ...customReducers })
  )
}

// Persist state.
export const saveState = (): void => {
  if (process.env.GATSBY_DISABLE_CACHE_PERSISTENCE) {
    // do not persist cache if above env var is set.
    // this is to temporarily unblock builds that hit the v8.serialize related
    // Node.js buffer size exceeding kMaxLength fatal crashes
    return undefined
  }

  const state = store.getState()

  const sliceOfStateToPersist = _.pick(state, persistedReduxKeys)

  return writeToCache(sliceOfStateToPersist)
}

export const savePartialStateToDisk = (
  slices: Array<GatsbyStateKeys>,
  optionalPrefix?: string,
  transformState?: <T extends DeepPartial<IGatsbyState>>(state: T) => T
): void => {
  const state = store.getState()
  const contents = _.pick(state, slices)
  const savedContents = transformState ? transformState(contents) : contents

  return writeToCache(savedContents, slices, optionalPrefix)
}

export const loadPartialStateFromDisk = (
  slices: Array<GatsbyStateKeys>,
  optionalPrefix?: string
): DeepPartial<IGatsbyState> => {
  try {
    return readFromCache(slices, optionalPrefix) as DeepPartial<IGatsbyState>
  } catch (e) {
    // ignore errors.
  }
  return {} as IGatsbyState
}

store.subscribe(() => {
  const lastAction = store.getState().lastAction
  emitter.emit(lastAction.type, lastAction)
})
