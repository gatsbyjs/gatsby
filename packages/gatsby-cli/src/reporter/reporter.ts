import { stripIndent } from "common-tags"
import chalk from "chalk"
import { globalTracer, Span, SpanContext } from "opentracing"

import * as reduxReporterActions from "./redux/actions"
import { LogLevels, ActivityStatuses } from "./constants"
import { getErrorFormatter } from "./errors"
import constructError from "../structured-errors/construct-error"
import {
  IErrorMapEntryPublicApi,
  ErrorId,
} from "../structured-errors/error-map"
import { prematureEnd } from "./catch-exit-signals"
import { IConstructError, IStructuredError } from "../structured-errors/types"
import { createTimerReporter, ITimerReporter } from "./reporter-timer"
import { createPhantomReporter, IPhantomReporter } from "./reporter-phantom"
import { createProgressReporter, IProgressReporter } from "./reporter-progress"
import {
  ErrorMeta,
  CreateLogAction,
  ILogIntent,
  IRenderPageArgs,
} from "./types"
import {
  registerAdditionalDiagnosticOutputHandler,
  AdditionalDiagnosticsOutputHandler,
} from "./redux/diagnostics"
import { isTruthy } from "gatsby-core-utils/is-truthy"

const errorFormatter = getErrorFormatter()
const tracer = globalTracer()

let reporterActions = reduxReporterActions

export interface IActivityArgs {
  id?: string
  parentSpan?: Span | SpanContext
  tags?: { [key: string]: any }
}

// eslint-disable-next-line prefer-const
let isVerbose = isTruthy(process.env.GATSBY_REPORTER_ISVERBOSE)

function isLogIntentMessage(msg: any): msg is ILogIntent {
  return msg && msg.type === `LOG_INTENT`
}

/**
 * Reporter module.
 * @module reporter
 */
class Reporter {
  /**
   * Strip initial indentation template function.
   */
  stripIndent = stripIndent
  format = chalk

  errorMap: Record<ErrorId, IErrorMapEntryPublicApi> = {}

  /**
   * Set a custom error map to the reporter. This allows
   * the reporter to extend the internal error map
   *
   * Please note: The entered IDs ideally should be different from the ones we internally use:
   * https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-cli/src/structured-errors/error-map.ts
   */

  setErrorMap = (entry: Record<string, IErrorMapEntryPublicApi>): void => {
    this.errorMap = {
      ...this.errorMap,
      ...entry,
    }
  }

  /**
   * Toggle verbosity.
   */
  setVerbose = (_isVerbose: boolean = true): void => {
    isVerbose = _isVerbose
    process.env.GATSBY_REPORTER_ISVERBOSE = isVerbose ? `1` : `0`
  }

  /**
   * Turn off colors in error output.
   */
  setNoColor = (isNoColor: boolean = false): void => {
    if (isNoColor) {
      errorFormatter.withoutColors()
    }

    // disables colors in popular terminal output coloring packages
    //  - chalk: see https://www.npmjs.com/package/chalk#chalksupportscolor
    //  - ansi-colors: see https://github.com/doowb/ansi-colors/blob/8024126c7115a0efb25a9a0e87bc5e29fd66831f/index.js#L5-L7
    if (isNoColor) {
      process.env.FORCE_COLOR = `0`
      // chalk determines color level at import time. Before we reach this point,
      // chalk was already imported, so we need to retroactively adjust level
      chalk.level = 0
    }
  }

  /**
   * Log arguments and exit process with status 1.
   */
  panic = (
    errorMeta: ErrorMeta,
    error?: Error | Array<Error>,
    pluginName?: string
  ): never => {
    this.error(errorMeta, error, pluginName)
    prematureEnd()
    return process.exit(1)
  }

  panicOnBuild = (
    errorMeta: ErrorMeta,
    error?: Error | Array<Error>,
    pluginName?: string
  ): IStructuredError | Array<IStructuredError> => {
    const reporterError = this.error(errorMeta, error, pluginName)
    if (process.env.gatsby_executing_command === `build`) {
      prematureEnd()
      process.exit(1)
    }
    return reporterError
  }

  error = (
    errorMeta: ErrorMeta | Array<ErrorMeta>,
    error?: Error | Array<Error>,
    pluginName?: string
  ): IStructuredError | Array<IStructuredError> => {
    let details: IConstructError["details"] = {
      context: {},
    }

    // Many paths to retain backcompat :scream:
    // 1.
    //   reporter.error(any, Error);
    //   reporter.error(any, [Error]);
    if (error) {
      if (Array.isArray(error)) {
        return error.map(errorItem =>
          this.error(errorMeta, errorItem)
        ) as Array<IStructuredError>
      }
      details.error = error
      details.context = {
        sourceMessage: errorMeta + ` ` + error.message,
      }
      // 2.
      //    reporter.error(Error);
    } else if (errorMeta instanceof Error) {
      details.error = errorMeta
      details.context = {
        sourceMessage: errorMeta.message,
      }
      // 3.
      //    reporter.error([Error]);
    } else if (Array.isArray(errorMeta)) {
      // when we get an array of messages, call this function once for each error
      return errorMeta.map(errorItem =>
        this.error(errorItem)
      ) as Array<IStructuredError>
      // 4.
      //    reporter.error(errorMeta);
    } else if (typeof errorMeta === `object`) {
      details = { ...errorMeta }
      // 5.
      //    reporter.error('foo');
    } else if (typeof errorMeta === `string`) {
      details.context = {
        sourceMessage: errorMeta,
      }
    }

    if (pluginName) {
      details.pluginName = pluginName
      const id = details?.id

      if (id) {
        const isPrefixed = id.includes(`${pluginName}_`)
        if (!isPrefixed) {
          details.id = `${pluginName}_${id}`
        }
      }
    }

    const structuredError = constructError({ details }, this.errorMap)

    if (structuredError) {
      reporterActions.createLog(structuredError)
    }

    // TODO: remove this once Error component can render this info
    // log formatted stacktrace
    if (structuredError.error) {
      this.log(errorFormatter.render(structuredError.error))
    }
    return structuredError
  }

  /**
   * Set prefix on uptime.
   */
  uptime = (prefix: string): void => {
    this.verbose(`${prefix}: ${(process.uptime() * 1000).toFixed(3)}ms`)
  }

  verbose = (text: string): void => {
    if (isVerbose) {
      reporterActions.createLog({
        level: LogLevels.Debug,
        text,
      })
    }
  }

  success = (text?: string): CreateLogAction =>
    reporterActions.createLog({ level: LogLevels.Success, text })
  info = (text?: string): CreateLogAction =>
    reporterActions.createLog({ level: LogLevels.Info, text })
  warn = (text?: string): CreateLogAction =>
    reporterActions.createLog({ level: LogLevels.Warning, text })
  log = (text?: string): CreateLogAction =>
    reporterActions.createLog({ level: LogLevels.Log, text })

  pendingActivity = reporterActions.createPendingActivity

  completeActivity = (
    id: string,
    status: ActivityStatuses = ActivityStatuses.Success
  ): void => {
    reporterActions.endActivity({ id, status })
  }

  /**
   * Time an activity.
   */
  activityTimer = (
    text: string,
    activityArgs: IActivityArgs = {},
    pluginName?: string
  ): ITimerReporter => {
    let { parentSpan, id, tags } = activityArgs
    const spanArgs = parentSpan ? { childOf: parentSpan, tags } : { tags }
    if (!id) {
      id = text
    }

    const span = tracer.startSpan(text, spanArgs)

    return createTimerReporter({
      text,
      id,
      span,
      reporter: this,
      reporterActions,
      pluginName,
    })
  }

  /**
   * Create an Activity that is not visible to the user
   *
   * During the lifecycle of the Gatsby process, sometimes we need to do some
   * async work and wait for it to complete. A typical example of this is a job.
   * This work should set the status of the process to `in progress` while running and
   * `complete` (or `failure`) when complete. Activities do just this! However, they
   * are visible to the user. So this function can be used to create a _hidden_ activity
   * that while not displayed in the CLI, still triggers a change in process status.
   */
  phantomActivity = (
    text: string,
    activityArgs: IActivityArgs = {}
  ): IPhantomReporter => {
    let { parentSpan, id, tags } = activityArgs
    const spanArgs = parentSpan ? { childOf: parentSpan, tags } : { tags }
    if (!id) {
      id = text
    }

    const span = tracer.startSpan(text, spanArgs)

    return createPhantomReporter({ id, text, span, reporterActions })
  }

  /**
   * Create a progress bar for an activity
   */
  createProgress = (
    text: string,
    total = 0,
    start = 0,
    activityArgs: IActivityArgs = {},
    pluginName?: string
  ): IProgressReporter => {
    let { parentSpan, id, tags } = activityArgs
    const spanArgs = parentSpan ? { childOf: parentSpan, tags } : { tags }
    if (!id) {
      id = text
    }
    const span = tracer.startSpan(text, spanArgs)

    return createProgressReporter({
      id,
      text,
      total,
      start,
      span,
      reporter: this,
      reporterActions,
      pluginName,
    })
  }

  // This method was called in older versions of gatsby, so we need to keep it to avoid
  // "reporter._setStage is not a function" error when gatsby@<2.16 is used with gatsby-cli@>=2.8
  _setStage = (): void => {}

  // This method is called by core when initializing worker process, so it can communicate with main process
  // and dispatch structured logs created by workers to parent process.
  _initReporterMessagingInWorker(sendMessage: (msg: ILogIntent) => void): void {
    const intentifiedActionCreators = {}
    for (const actionCreatorName of Object.keys(reduxReporterActions) as Array<
      keyof typeof reduxReporterActions
    >) {
      // swap each reporter action creator with function that send intent
      // to main process
      intentifiedActionCreators[actionCreatorName] = (...args): void => {
        sendMessage({
          type: `LOG_INTENT`,
          payload: {
            name: actionCreatorName,
            args,
          } as any,
        })
      }
    }
    reporterActions = intentifiedActionCreators as typeof reduxReporterActions
  }

  // This method is called by core when initializing worker pool, so main process can receive
  // messages from workers and dispatch structured logs created by workers to parent process.
  _initReporterMessagingInMain(
    onMessage: (listener: (msg: ILogIntent | unknown) => void) => void
  ): void {
    onMessage(msg => {
      if (isLogIntentMessage(msg)) {
        reduxReporterActions[msg.payload.name].call(
          reduxReporterActions,
          // @ts-ignore Next line (`...msg.payload.args`) cause "A spread argument
          // must either have a tuple type or be passed to a rest parameter"
          ...msg.payload.args
        )
      }
    })
  }

  _renderPageTree(args: IRenderPageArgs): void {
    reporterActions.renderPageTree(args)
  }

  _registerAdditionalDiagnosticOutputHandler(
    handler: AdditionalDiagnosticsOutputHandler
  ): void {
    registerAdditionalDiagnosticOutputHandler(handler)
  }
}
export type { Reporter }
export const reporter = new Reporter()
