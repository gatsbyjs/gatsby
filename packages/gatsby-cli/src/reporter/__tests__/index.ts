import { Level } from "../../structured-errors/types"
import { reporter } from "../reporter"
import * as reporterActions from "../redux/actions"

// TODO: report.error now DOES return something. Get rid of this spying mocking stuff

// report.error doesn't return anything, it creates a `structuredError` object and
// calls reporterInstance.error(structuredError)

// Spy on reporterInstance.error and mock its implementation so it
// returns structuredError

// We can then use the returned structuredError for snapshots.
jest
  .spyOn(reporterActions, `createLog`)
  // @ts-ignore
  .mockImplementation(structuredLog => structuredLog)

// We don't care about this
reporter.log = jest.fn()

const getErrorMessages = (fn: jest.Mock): any =>
  fn.mock.calls
    .map(([firstArg]) => firstArg)
    .filter(structuredMessage => structuredMessage.level === `ERROR`)

describe(`report.error`, () => {
  beforeEach(() => {
    ;(reporterActions.createLog as jest.Mock).mockClear()
  })

  it(`handles "String, Error" signature correctly`, () => {
    reporter.error(
      `Error string passed to reporter`,
      new Error(`Message from new Error`)
    )
    const generatedError = getErrorMessages(
      reporterActions.createLog as jest.Mock
    )[0]

    expect(generatedError).toMatchSnapshot({
      stack: expect.any(Array),
    })
  })

  it(`handles "Error" signature correctly`, () => {
    reporter.error(new Error(`Message from new Error`))
    const generatedError = getErrorMessages(
      reporterActions.createLog as jest.Mock
    )[0]
    expect(generatedError).toMatchSnapshot({
      stack: expect.any(Array),
    })
  })

  it(`handles "Array of Errors" signature correctly`, () => {
    reporter.error([
      new Error(`Message 1 from new Error`),
      new Error(`Message 2 from new Error`),
      new Error(`Message 3 from new Error`),
    ])

    const generatedErrors = getErrorMessages(
      reporterActions.createLog as jest.Mock
    )

    expect(generatedErrors.length).toEqual(3)

    // get final generated object
    const generatedError = generatedErrors[2]
    expect(generatedError).toMatchSnapshot({
      stack: expect.any(Array),
    })
  })

  it(`handles "structuredError" signature correctly`, () => {
    reporter.error({
      id: `95312`,
      context: {
        undefinedGlobal: `navigator`,
      },
    })
    const generatedError = getErrorMessages(
      reporterActions.createLog as jest.Mock
    )[0]
    expect(generatedError).toMatchSnapshot({
      stack: expect.any(Array),
    })
  })

  it(`handles "String" signature correctly`, () => {
    reporter.error(`Error created in Jest`)
    const generatedError = getErrorMessages(
      reporterActions.createLog as jest.Mock
    )[0]
    expect(generatedError).toMatchSnapshot({
      stack: expect.any(Array),
    })
  })

  it(`handles "String, Error, pluginName" signature correctly`, () => {
    reporter.error(
      `Error string passed to reporter`,
      new Error(`Message from new Error`),
      `gatsby-plugin-foo-bar`
    )
    const generatedError = getErrorMessages(
      reporterActions.createLog as jest.Mock
    )[0]
    expect(generatedError).toMatchSnapshot({
      stack: expect.any(Array),
    })
  })

  it(`sets an error map if setErrorMap is called`, () => {
    reporter.setErrorMap({
      "1337": {
        text: (context): string => `Error text is ${context.someProp}`,
        level: Level.ERROR,
        docsUrl: `https://www.gatsbyjs.com/docs/gatsby-cli/#new`,
      },
    })

    expect(reporter.errorMap[`1337`]).toBeTruthy()
  })

  it(`uses custom error from errorMap`, () => {
    reporter.setErrorMap({
      "1337": {
        text: (context): string => `Error text is ${context.someProp}`,
        level: Level.ERROR,
        docsUrl: `https://www.gatsbyjs.com/docs/gatsby-cli/#new`,
      },
    })

    reporter.error({
      id: `1337`,
      context: {
        someProp: `test123`,
      },
    })
    const generatedError = getErrorMessages(
      reporterActions.createLog as jest.Mock
    )[0]
    expect(generatedError).toMatchSnapshot({
      stack: expect.any(Array),
    })
  })

  // This is how it's potentially called from api-runner-node.js
  // It'll prefix the errorMap and then pass the pluginName as third arg
  it(`uses custom error from errorMap with pluginName`, () => {
    reporter.setErrorMap({
      "gatsby-plugin-foo-bar_1337": {
        text: (context): string => `Error text is ${context.someProp}`,
        level: Level.ERROR,
        docsUrl: `https://www.gatsbyjs.com/docs/gatsby-cli/#new`,
      },
    })

    reporter.error(
      {
        id: `1337`,
        context: {
          someProp: `test123`,
        },
      },
      undefined,
      `gatsby-plugin-foo-bar`
    )
    const generatedError = getErrorMessages(
      reporterActions.createLog as jest.Mock
    )[0]
    expect(generatedError).toMatchSnapshot({
      stack: expect.any(Array),
    })
  })
})
