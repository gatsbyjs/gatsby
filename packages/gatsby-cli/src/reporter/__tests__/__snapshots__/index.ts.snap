// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report.error handles "Array of Errors" signature correctly 1`] = `
Object {
  "category": "UNKNOWN",
  "context": Object {
    "sourceMessage": "Message 3 from new Error",
  },
  "docsUrl": "https://gatsby.dev/issue-how-to",
  "error": [Error: Message 3 from new Error],
  "level": "ERROR",
  "stack": Any<Array>,
  "text": "Message 3 from new Error",
  "type": "UNKNOWN",
}
`;

exports[`report.error handles "Error" signature correctly 1`] = `
Object {
  "category": "UNKNOWN",
  "context": Object {
    "sourceMessage": "Message from new Error",
  },
  "docsUrl": "https://gatsby.dev/issue-how-to",
  "error": [Error: Message from new Error],
  "level": "ERROR",
  "stack": Any<Array>,
  "text": "Message from new Error",
  "type": "UNKNOWN",
}
`;

exports[`report.error handles "String" signature correctly 1`] = `
Object {
  "category": "UNKNOWN",
  "context": Object {
    "sourceMessage": "Error created in Jest",
  },
  "docsUrl": "https://gatsby.dev/issue-how-to",
  "level": "ERROR",
  "stack": Any<Array>,
  "text": "Error created in Jest",
  "type": "UNKNOWN",
}
`;

exports[`report.error handles "String, Error" signature correctly 1`] = `
Object {
  "category": "UNKNOWN",
  "context": Object {
    "sourceMessage": "Error string passed to reporter Message from new Error",
  },
  "docsUrl": "https://gatsby.dev/issue-how-to",
  "error": [Error: Message from new Error],
  "level": "ERROR",
  "stack": Any<Array>,
  "text": "Error string passed to reporter Message from new Error",
  "type": "UNKNOWN",
}
`;

exports[`report.error handles "String, Error, pluginName" signature correctly 1`] = `
Object {
  "category": "UNKNOWN",
  "context": Object {
    "sourceMessage": "Error string passed to reporter Message from new Error",
  },
  "docsUrl": "https://gatsby.dev/issue-how-to",
  "error": [Error: Message from new Error],
  "level": "ERROR",
  "pluginName": "gatsby-plugin-foo-bar",
  "stack": Any<Array>,
  "text": "Error string passed to reporter Message from new Error",
  "type": "UNKNOWN",
}
`;

exports[`report.error handles "structuredError" signature correctly 1`] = `
Object {
  "category": "USER",
  "code": "95312",
  "context": Object {
    "undefinedGlobal": "navigator",
  },
  "docsUrl": "https://gatsby.dev/debug-html",
  "level": "ERROR",
  "stack": Any<Array>,
  "text": "\\"navigator\\" is not available during server-side rendering. Enable \\"DEV_SSR\\" to debug this during \\"gatsby develop\\".",
  "type": "HTML.COMPILATION",
}
`;

exports[`report.error uses custom error from errorMap 1`] = `
Object {
  "code": "1337",
  "context": Object {
    "someProp": "test123",
  },
  "docsUrl": "https://www.gatsbyjs.com/docs/gatsby-cli/#new",
  "level": "ERROR",
  "stack": Any<Array>,
  "text": "Error text is test123",
  "type": "PLUGIN",
}
`;

exports[`report.error uses custom error from errorMap with pluginName 1`] = `
Object {
  "code": "gatsby-plugin-foo-bar_1337",
  "context": Object {
    "someProp": "test123",
  },
  "docsUrl": "https://www.gatsbyjs.com/docs/gatsby-cli/#new",
  "level": "ERROR",
  "pluginName": "gatsby-plugin-foo-bar",
  "stack": Any<Array>,
  "text": "Error text is test123",
  "type": "PLUGIN",
}
`;
